File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_3_cross_cutting_concerns/before/OrderService.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_3_cross_cutting_concerns.before;

/**
 * BUSINESS LOGIC LAYER.
 * ARCHITECTURE WARNING: This class is "welded" to the StaticFileLogger.
 * You cannot test SaveOrder without also executing the static
 * logger logic. This violates the Dependency Inversion Principle.
 */
public class OrderService {
    public void saveOrder(Order order) {
        // HIDDEN DEPENDENCY: This is not visible in the API.
        System.out.println("Saving order..." + order.getId());
        System.out.println("(BEFORE_SERVICE) Order saved.");
    }
}

File: Chapter 04/JS/section_4_3_cross_cutting_concerns/before/orderService.js
const StaticFileLogger = require('./staticFileLogger');

/**
 * ARCHITECTURE WARNING: Rigid Dependency.
 * The service is tightly coupled to the specific logger 
 * implementation. Swapping this for a cloud logger would 
 * require modifying this file.
 */
class OrderService {
    saveOrder(order) {
        // 🚨 VIOLATION: Hardcoded reference to a static utility.
        StaticFileLogger.log("Saving order...");
        console.log("(BEFORE_SERVICE) Order saved.");
    }
}

module.exports = OrderService;